### ISSUE TITLE
How to further optimize model for single image faster inference

### ISSUE THREAD

**Issue by EmmanuelMess** [ultralytics/ultralytics#18864]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am working on detecting faces on a Raspberry Pi Zero. For this I'm using a YOLOv6 modified to reduce the number of parameters (removed heads, reduced image size), and exporting a quantized version ("best_integer_quant.tflite"), but it is surprisingly slow in the Pi (~800ms best case), are there any strategies I'm missing to make it run faster?

Training:
```
model = YOLO("./yolov6-face.yaml")
r = model.train(data="face-detection-dataset.yaml", epochs=50, imgsz='192,320', single_cls=True, plots=True, batch=0.9)
```

model yaml:

```
%%writefile yolov6-face.yaml
# Ultralytics YOLO 🚀, AGPL-3.0 license
# YOLOv6 object detection model with P3-P5 outputs. For Usage examples see https://docs.ultralytics.com/models/yolov6

# Parameters
nc: 1 # number of classes
activation: nn.ReLU() # (optional) model default activation function
scales: # model compound scaling constants, i.e. 'model=yolov6n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  p: [0.33, 0.25, 8] # nano is [0.33, 0.25, 1024]

# YOLOv6-3.0s backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2/4
  - [-1, 6, Conv, [128, 3, 1]]
  - [-1, 1, Conv, [256, 3, 2]] # 3-P3/8
  - [-1, 12, Conv, [256, 3, 1]]
  - [-1, 1, Conv, [512, 3, 2]] # 5-P4/16
  - [-1, 18, Conv, [512, 3, 1]]
  - [-1, 1, Conv, [1024, 3, 2]] # 7-P5/32
  - [-1, 6, Conv, [1024, 3, 1]]
  - [-1, 1, SPPF, [1024, 5]] # 9

# YOLOv6-3.0s head
head:
  - [-1, 1, Conv, [256, 1, 1]]
  - [-1, 1, nn.ConvTranspose2d, [256, 2, 2, 0]]
  - [[-1, 6], 1, Concat, [1]] # cat backbone P4
  - [-1, 1, Conv, [256, 3, 1]]
  - [-1, 9, Conv, [256, 3, 1]] # 14

  - [-1, 1, Conv, [128, 1, 1]]
  - [-1, 1, nn.ConvTranspose2d, [128, 2, 2, 0]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 1, Conv, [128, 3, 1]]
  - [-1, 9, Conv, [128, 3, 1]] # 19

  - [[14, 19], 1, Detect, [nc]] # Detect(P3, P4, P5)
```

Model printout:
```
                   from  n    params  module                                       arguments                     
  0                  -1  1       232  ultralytics.nn.modules.conv.Conv             [3, 8, 3, 2]                  
  1                  -1  1       592  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 2]                  
  2                  -1  2      1184  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 1]                  
  3                  -1  1       592  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 2]                  
  4                  -1  4      2368  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 1]                  
  5                  -1  1       592  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 2]                  
  6                  -1  6      3552  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 1]                  
  7                  -1  1       592  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 2]                  
  8                  -1  2      1184  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 1]                  
  9                  -1  1       184  ultralytics.nn.modules.block.SPPF            [8, 8, 5]                     
 10                  -1  1        80  ultralytics.nn.modules.conv.Conv             [8, 8, 1, 1]                  
 11                  -1  1       264  torch.nn.modules.conv.ConvTranspose2d        [8, 8, 2, 2, 0]               
 12             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 13                  -1  1      1168  ultralytics.nn.modules.conv.Conv             [16, 8, 3, 1]                 
 14                  -1  3      1776  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 1]                  
 15                  -1  1        80  ultralytics.nn.modules.conv.Conv             [8, 8, 1, 1]                  
 16                  -1  1       264  torch.nn.modules.conv.ConvTranspose2d        [8, 8, 2, 2, 0]               
 17             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1      1168  ultralytics.nn.modules.conv.Conv             [16, 8, 3, 1]                 
 19                  -1  3      1776  ultralytics.nn.modules.conv.Conv             [8, 8, 3, 1]                  
 20            [14, 19]  1     94178  ultralytics.nn.modules.head.Detect           [1, [8, 8]]                   
YOLOv6-face summary: 145 layers, 111,826 parameters, 111,810 gradients, 1.0 GFLOPs
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @EmmanuelMess, thank you for your interest in Ultralytics 🚀! We appreciate you sharing your detailed setup and question. Here's how we can assist further:

We recommend reviewing the [Ultralytics Docs](https://docs.ultralytics.com/) for helpful resources on model usage, optimization, and deployment. Additionally:

---

## ❓ If this is a custom optimization question:
Please provide detailed context to assist further. It seems like you're encountering performance bottlenecks on a Raspberry Pi Zero. If you haven't already, consider the following strategies:
- Experiment with smaller model architectures like YOLOv5n for ultralight setups.
- Check the efficiency of your environment (Python version, TensorFlow Lite setup, etc.).
- Explore hardware-assisted frameworks like Coral or NVIDIA TensorRT (dependant on your hardware).

You can explore additional performance optimization tips in our [Guides](https://docs.ultralytics.com/guides/). 

## 🐛 If this is a bug report:
Please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that replicates the issue you're experiencing to allow us to assist more effectively. This includes:
- A concise code snippet outlining the bug
- Any specific errors or warnings you're encountering
- Logs from your deployment environment (if applicable)

---

## Upgrade

Ensure you're using the latest version of the `ultralytics` package with all dependencies updated to avoid compatibility issues:
```bash
pip install -U ultralytics
```

---

## Deployment Environments 🌐

YOLO models can be run in several environments with pre-installed dependencies, optimized for different hardware setups:
- **Notebooks** with free GPUs: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** - [GCP Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon AWS** - [AWS Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker** - [Docker Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

---

## Community and Support 🤝

Join and stay connected with the Ultralytics community:
- **Discord**: [Join real-time chat](https://discord.com/invite/ultralytics) 🎧
- **Discourse Forums**: [Explore in-depth discussions](https://community.ultralytics.com/)
- **Subreddit**: [Join discussions on Reddit](https://reddit.com/r/Ultralytics)

---

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If the CI badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests validate YOLO functionality across macOS, Windows, and Ubuntu every 24 hours.

---

This is an automated response to help guide you to solutions faster. An Ultralytics engineer will review your issue soon and provide further insights 😊

**Comment by glenn-jocher** [comment]
@EmmanuelMess to optimize inference speed for your Raspberry Pi Zero, consider these strategies:

1. **INT8 Quantization Optimization**: Ensure proper calibration during quantization to maintain performance. Use TensorRT INT8 if supported, as it can offer significant speed improvements. Refer to the [TensorRT Integration](https://docs.ultralytics.com/integrations/tensorrt/) for details.

2. **Model Complexity**: Further reduce model complexity by pruning layers, reducing input resolution, or employing a more lightweight architecture like YOLOv5n or YOLOv8n, which are optimized for edge devices.

3. **Batch Size**: Confirm that the batch size is set to 1 during inference to minimize resource usage.

4. **Hardware-Specific Optimizations**: Explore libraries like Coral Edge TPU or OpenVINO if the Raspberry Pi Zero supports such accelerators. These tools can optimize inference for specific hardware.

5. **Efficient Backend**: Use a highly optimized runtime like ONNX Runtime or TFLite with hardware acceleration enabled.

For consistent inference times (e.g., 0.3ms on P100), this might result from CUDA-level optimizations or static graph compilation, irrespective of model complexity. This is typical on high-end hardware and does not necessarily translate to resource-constrained devices like Raspberry Pi.

Further guidance is available in our [Benchmarking documentation](https://docs.ultralytics.com/modes/benchmark/).

**Comment by Y-T-G** [comment]
NCNN works better with RPI

**Comment by EmmanuelMess** [comment]
Benchmark is having some issues for me, I've added created #18883 for those. I will try to compile NCNN for RPi Zero.

**Comment by glenn-jocher** [comment]
Thank you for raising #18883; the team will review it. Compiling NCNN for the RPi Zero is a great approach, as it is optimized for resource-constrained devices. Let us know if you encounter further issues.

**Comment by bshakya77** [comment]
@ Hello @glenn-jocher,

I have trained YOLOv11n model in DOTA1.5 dataset. I am trying to report the inference time in milliseconds. I find two difference inference time during validation task and prediction task. The snippets of both tasks inference are attached. I would like to know which is the accurate inference time to be considered for reporting.

![Image](https://github.com/user-attachments/assets/5c48865e-9b2f-49e4-a1fc-7ce308ce2be5)
![Image](https://github.com/user-attachments/assets/1ac37ab7-4946-4524-a220-e5fd761d33ca)

I appreciate your feedback.

Regards,
Bijay

**Comment by glenn-jocher** [comment]
Hi Bijay,  
  
For reporting inference times, use the **prediction** task metrics (preprocess + inference + postprocess) as they reflect real-world deployment performance. Validation includes additional metric calculations that aren't part of runtime inference. For standardized benchmarking, see our [Benchmarking Documentation](https://docs.ultralytics.com/modes/benchmark/).  

Best regards

**Comment by bshakya77** [comment]
Hello @glenn-jocher,

Thank you for your clarification. 

Regards,
Bijay

**Comment by glenn-jocher** [comment]
You're welcome! Let us know if you have any other YOLO-related questions.

**Comment by bshakya77** [comment]
Hello @glenn-jocher,

I would like to know if the `ultralytics` provides the validation and prediction for YOLOX model in the same way as other YOLO variants. If you have any references documentations, it would be helpful.

Thank you.

Regards,
Bijay

**Comment by glenn-jocher** [comment]
Hi Bijay,  
  
Yes, Ultralytics YOLOX models support validation and prediction just like other YOLO variants. For details, see our [YOLOv8 vs YOLOX comparison guide](https://docs.ultralytics.com/compare/yolov8-vs-yolox/), which includes usage examples and performance benchmarks. 

Best regards

**Comment by EmmanuelMess** [comment]
For lack of time I couldn't test NCNN, we ended up using tflite, and it works more or less as it needs to.

**Comment by glenn-jocher** [comment]
@EmmanuelMess thanks for the update! Glad to hear TFLite is working for your needs. If you revisit optimization later, our [Benchmarking documentation](https://docs.ultralytics.com/modes/benchmark/) offers insights into different deployment backends. Let us know if you need further guidance!
